torchgeo/datasets/seasonet.py

```diff
@@ -0,0 +1,456 @@
+    def _load_image(self, index: int) -> Tensor:
+        """Load image(s) for a single location.
+
+        Args:
+            index: index to return
+
+        Returns:
+            the stacked seasonal images
+        """
+        paths = self.files.iloc[index]
+        if self.concat_seasons > 1:
+            paths = random.sample(paths, self.concat_seasons)
+        tensor = torch.empty(self.concat_seasons * self.channels, *self.image_size)
+        for img_idx, path in enumerate(paths):
+            bnd_idx = 0
+            for band in self.bands:
+                with rasterio.open(f"{path}_{band}.tif") as f:
+                    array = f.read(
+                        out_shape=[f.count] + list(self.image_size),
+                        out_dtype="int32",
+                        resampling=Resampling.bilinear,
+                    )
+                image = torch.from_numpy(array).float()
+                c = img_idx * self.channels + bnd_idx
+                tensor[c : c + image.shape[0]] = image
+                bnd_idx += image.shape[0]
+        return tensor
+
+    def _load_target(self, index: int) -> Tensor:
+        """Load the target mask for a single location.
+
+        Args:
+            index: index to return
+
+        Returns:
+            the target mask
+        """
+        path = self.files.iloc[index][0]
+        with rasterio.open(f"{path}_labels.tif") as f:
+            array = f.read() - 1
+        tensor = torch.from_numpy(array).squeeze().long()
+        return tensor
+
+    def _verify(self) -> None:
+        """Verify the integrity of the dataset.
+
+        Raises:
+            RuntimeError: if ``download=False`` but dataset is missing or checksum fails
+        """
+        # Check if all files already exist
+        if all(
+            os.path.exists(os.path.join(self.root, name))
+            for name in self.extracted_names
+        ):
+            return
+
+        # Check for downloaded files and extract existing zips
+        missing = []
+        for file_name in self.file_names:
+            file_path = os.path.join(self.root, file_name)
+            if not os.path.exists(file_path):
+                missing.append(file_name)
+            elif file_name[-3:] == "zip":
+                extract_archive(file_path)
+
+        if not missing:
+            return
+
+        # Check if the user requested to download the dataset
+        if not self.download:
+            missing_str = (
+                f"{', '.join(missing[:-1])} and {missing[-1]}"
+                if len(missing) > 1
+                else missing[0]
+            )
+            raise RuntimeError(
+                f"{missing_str} not found in `root={self.root}` and `download=False`, "
+                "either specify a different `root` directory or use `download=True`"
+                " to automatically download the dataset."
+            )
+
+        # Download missing files
+        record_info = requests.get(self.url).json()["files"]
```

Comment:
Is this necessary? None of our other datasets directly use requests to download, they rely on download_url.

Reply:
download_url wouldn't work in this case, because no files are being downloaded here. This is just a call to the Zenodo API to get the correct file URLs and MD5s and is also the intended way to download datasets from Zenodo through scripts. This is also what the MockResponse class in our tests is for, as it mocks the response to this API call.

Technically it would be possible to hard-code all URLs and MD5s into the dataset but in case anything ever changes either on our side regarding the files themselves or on Zenodos side regarding their location, this API call would still provide the correct information.

Reply:
How many URLs/MD5s are we talking? The whole point of MD5s is to ensure that the files that get downloaded are the ones expected. For example, if someone hacked your Zenodo account and uploaded a virus in place of the dataset, a change in MD5 would alert users to the issue. I don't think our users are hugely concerned about security, but MD5s are still useful to ensure reproducibility.

Reply:
That's basically the approach our current datasets use. The URL and MD5 are hard-coded into the dataset, and we optionally verify the checksum after download. Let's do the same for this dataset for now. If we're concerned that the URL might change frequently, we can do something fancier later. And the MD5 changing is less of a big deal since users can just use `checksum=False` if they don't care, which is the default.

Reply:
Afaik the approach most datasets use currently does not include an MD5 check before downloading, which I still think would be nice to have, since it would inform the user that something is wrong before having to download a 60GB file and computing the MD5 on their own machine. But just to make sure, since this is all concerning a very simple API request, just using urllib instead of requests should work, right? Since that is what `download_url` also uses.

Reply:
I'm not sure if most websites (other than Zenodo) will let you verify the checksum before you even start the download. For now, it's best to keep things the same as every other dataset. If you want to change how all dataset downloads work, that should be left up to a follow-up PR. Any code unique to this PR requires a lot more careful thought and vetting, whereas using the same code as every other dataset makes it easier for me to tell it will work consistently.